### ISSUE TITLE
Unlisted sub-dependencies don't resolve with parent config

### ISSUE THREAD

**Issue by Bajix** [stealjs/steal#971]
Here's a real world case I ran into:

jQuery UI doesn't have jQuery listed as a dependency, yet requires jQuery. The dependencies are then resolved incorrectly, failing to resolve using parent configuration. 

System._traceData.loads["jquery-ui@1.12.1#ui/widgets/autocomplete"].metadata.dependencies is resolving to "jquery" instead of "jquery@3.1.1/dist/jquery"

This problem can't be fixed with meta or map, as those aren't aware that jQuery is a npm dependency when tracing from within a package missing dependencies.

**Comment by m-mujica** [comment]
I wrote the following code to convert the `meta.deps` names through `npm-convert`

```javascript
/**
 * Convert each module name defined in `meta.deps`
 */
function convertMetaDeps(opts) {
  if (!opts.meta) {
    return;
  }

  for (var key in opts.meta) {
    var mod = opts.meta[key];

    if (mod.deps) {
      var converted = [];

      mod.deps.forEach(function(dep) {
        if (dep) {
          converted.push(convertName(opts.context, opts.pkg, opts.meta,
            opts.root, dep, opts.waiting));
        }
      });

      mod.deps = converted;
    }
  }
}
```

1) `convertName` returns `null` when the name conversion is queue because the parent package.json has not been crawled (probably bad idea to replace the `meta.deps` with `null` values).

2) The enqueued conversion does not get processed and the test keeps failing (ref: https://github.com/stealjs/steal/pull/1110)

@matthewp thoughts?

**Comment by matthewp** [comment]
I think contextual map configuration is likely the solution here. What we essentially have is that `foo` depends on `bar`, but `bar` doesn't mean anything in the **context** of foo. Contextual mapping is a way to configure contextual meanings:

```json
{
  "steal": {
    "map": {
      "foo": {
        "bar": "something-else"
      }
    }
  }
}
```

**Comment by matthewp** [comment]
Ok, this bears some explanation so please bear with me. The problem is unrelated to meta.deps. It's a more fundamental problem than that. So I'll use an example with CommonJS to show the problem more directly.

# Problem

**foo**

```js
require("bar");
```

Essentially what meta.deps does is "inject" a require call inside of a module (not really but essentially the same thing).

Normalization is contextual. The meaning of **bar** is in context of the parent, **foo**.

Since **bar** doesn't mean anything to **foo** it doesn't normalize correctly. What we are asking for is to *change the contextual meaning of bar*.

# Contextual maps

Luckily there is an existing way to do that, contextual maps.

```json
{
  "steal": {
    "map": {
      "foo": {
        "bar": "baz"
      }
    }
  }
}
```

This is a contextual map that says *when foo asks for bar, give it baz instead*.

# Another problem

This is a great elegant solution, the only problem is that `baz` here needs to be a moduleName, not a module identifier. That is, we cannot identify the meaning of `baz` in the context of `foo`. We need to identify it in the context of the package name that defines the configuration.

# Possible solution

Essentially what need is a way to declare that when we see `bar` as the name and `foo` as the parentName, use something else (like `app`) as the parent name instead.

**Comment by justinbmeyer** [comment]
The last 2 sections seem to contradict each other.  It seems to imply you like using contextual maps to fix this problem, but also don't.

**Comment by justinbmeyer** [comment]
Say JustifiedGallary did `require("jquery")` but it didn't list jQuery as a dependency.  Would steal work?  This would work in an `npm` context as long as some parent had `jquery` as a dependency.

This makes me think that contextual maps are not the solution.  The solution is to "walk" parent package.json's until you find one with `jquery` and use that.

**Comment by matthewp** [comment]
> The last 2 sections seem to contradict each other. It seems to imply you like using contextual maps to fix this problem, but also don't.

Contextual map don't fix the problem, because the right-hand side of the map needs to be a moduleName and we don't have that in this case.

> This makes me think that contextual maps are not the solution. The solution is to "walk" parent package.json's until you find one with jquery and use that.

Node doesn't walk up package.json, it walks up node_modules folders. We can't easily do that without causing a lot of 404s. We would have to ban the use of any bare module identifiers that aren't npm package names when using the npm plugin.

Since this would break existing tests we can't do that until 2.0 anyways.

**Comment by matthewp** [comment]
I'm not sure that banning bare specifiers that aren't npm packages is doable at all. Bare specifiers are used for `map` configuration quite a bit. [steal-css defines a $css](https://github.com/stealjs/steal-css/blob/4ddd7d73265d099da64f87d844211acb7f7ae114/package.json#L42). We also have a few special bare specifier names in steal like `@loader` and `@steal`.

**Comment by matthewp** [comment]
I still think the solution here is some type of configuration. Perhaps we'll keep it private. Something like:

```js
steal.config({
  context: {
    "jquery": {
      "justifiedGallery": "app@1.0.0#main"
    }
  }
});
```

This would look up `jquery` within the context of `app@1.0.0#main` instead of `justifiedGallery`.

**Comment by matthewp** [comment]
Getting back to the walking up package.json idea @justinbmeyer. We can't walk up node_modules like Node does, but perhaps we can walk up package.jsons that have already been fetched. Since the parent package.json contains the dependency we need, and it has already been loading, we might be able to find it that way.

**Comment by matthewp** [comment]
To make this work I think we need some type of new data structure that maps package names to all parents. Something like:

```js
jquery: {
  parentA: {},
  parentB: {}
}
```

This way if we are unable to find an identifier the normal way (like looking at the parent package's own package.json) we can just find it *somewhere*.

**Comment by justinbmeyer** [comment]
@matthewp I thought 0.16 (or maybe earlier) was doing this already.  At least when we had npm `2`, we knew the paths to every package.json:



```
node_modules/project-using-justified/package.json
node_modules/project-using-justified/node_modules/justified/package.json
```

If we had this structure ... all you'd have to do is "walk up" to your parent's `package.json` path ... then look in its package.json.

**Comment by matthewp** [comment]
The folder structure is not all that useful in npm3+. `justified` is most likely a sibling of `project-using-justified`. So you can't infer parent-child relationship based on node_modules/ nesting.

```
node_modules/project-using-justified/package.json
node_modules/justified/package.json
```

**Comment by Bajix** [comment]
What about simply being more reliant on a server-side build to complement? Configuration could be pre-computed and certainly there are other use cases - steal [eyeglass](https://www.npmjs.com/package/eyeglass) perhaps?

**Comment by matthewp** [comment]
@Bajix I'm very interested in pre-computed configuration in that way. Especially as we look to make Steal more compatible with `<script type=module>`.

**Comment by justinbmeyer** [comment]
@Bajix I however and pretty turned off by requiring server-side configuration, especially when it's not need (as is the case here).

**Comment by Bajix** [comment]
@justinbmeyer Sure, but if we rip off the band aid and require server-side builds, we facilitate a LOT of use cases that simply couldn't be done otherwise. There's a greater good case to be made

**Comment by justinbmeyer** [comment]
@Bajix I don't think so.  There are plenty of other module loaders if that's important to someone.  Imo, steal is philosophical about the client running the show as much as possible.

**Comment by justinbmeyer** [comment]
> The folder structure is not all that useful in npm3+. 

I think it would be useful if `jquery` has been loaded.  Then you could just look for a jQuery project in any parent.  This is what `npm2` would do anyway.  I think this would be a nice "fast-path" solution.  

>  new data structure that maps package names to all parents

Would you use that to look up the current package's parents?  I think that's the right way of doing it.  It's possible no other package has loaded jQuery yet.  


Say there are two packages using `project-using-justified` (`a` and `b`), each has a different version of jQuery so things might get installed like:

```
node_modules/
  a/
  b/
    node_modules/jquery/
  jquery/
  justifiedG/
```

Presumably here, finding the right `jquery` for `justifiedG` would have to use that data structure to figure out which packages load `justifiedG`.  It would need to check if they have a `jquery` dependency.  If they both did, it might need to be smart enough to use the one at the right path.

**Comment by Bajix** [comment]
@justinbmeyer To an extent, but if there were a happy medium, we could circumvent the insurmountable limitations that exist with this being fully client oriented. There is no technically viable client-side solution that will ever be comparable to (eyeglass)[https://www.npmjs.com/package/eyeglass]. A seamless integration with eyeglass would truly be a game changer as it opens up a lot of possibilities for configurable / themed SASS components. SASS is overdue for proper dependency management.

**Comment by matthewp** [comment]
I think Steal can be client-oriented while also having optimizations that are aided by cli tooling. We did this recently when we added [development bundles](http://stealjs.com/docs/StealJS.guides.development_bundles.html) that rely on running a cli tool.

Anyways, as this is an issue that needs to be fixed (and will be closed soon), I think it's best to take this discussion over to https://github.com/stealjs/rfcs. Just create and issue to kick off discussion. I'm happy to explore this path with you.
